### PR TITLE
feat: add leader election options

### PR DESCRIPTION
**Pull request by CharlesQQ**
Signed-off-by: charlesQQ <charles_ali@qq.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
allow user change leader election action

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```
`karmada-controller-manager`/`karmada-agent`: Introduced `--leader-elect-lease-duration`, `--leader-elect-renew-deadline` and `--leader-elect-retry-period` flags to specify leader election behavious.
```

### CODE REVIEW

**Comment by CharlesQQ** [comment]
@XiShanYongYe-Chang  help, can you figure out where is problem with e2e test?

**Comment by XiShanYongYe-Chang** [comment]
> @XiShanYongYe-Chang help, can you figure out where is problem with e2e test?

Error log: `cluster member1 not ready`, just need to have a retry.

**Comment by RainbowMango** [comment]
Re-triggered.

@CharlesQQ I'm ok to add these flags, can you tell me why you need them. I'm not clear on the use cases.

**Comment by CharlesQQ** [comment]
> I'm not clear on the use cases.
leader election package fetch electionId (such as `configmap or leases.coordination.k8s.io`) from apiserver every 2 seconds, and retry 5 times; after 10 seconds, if still failed, controller process will exit; in some scenarios, use might want fetch slowly or more retry times, so that process will restart slowly

**Comment by RainbowMango** [comment]
Do you mean the default [renew deadline](https://github.com/kubernetes-sigs/controller-runtime/blob/ffd9ec8768b7d2e82aee6b0c76d481a5bdda4075/pkg/manager/manager.go#L202) (defaults to 10s) and [RetryPeriod](https://github.com/kubernetes-sigs/controller-runtime/blob/ffd9ec8768b7d2e82aee6b0c76d481a5bdda4075/pkg/manager/manager.go#L205) (defaults to 2s) might be too short for you, right?

File: cmd/agent/app/options/options.go [hunk 1]
```diff
@@ -98,6 +98,19 @@ func (o *Options) AddFlags(fs *pflag.FlagSet, allControllers []string) {
 	))
 	fs.BoolVar(&o.LeaderElection.LeaderElect, "leader-elect", true, "Start a leader election client and gain leadership before executing the main loop. Enable this when running replicated components for high availability.")
 	fs.StringVar(&o.LeaderElection.ResourceNamespace, "leader-elect-resource-namespace", util.NamespaceKarmadaSystem, "The namespace of resource object that is used for locking during leader election.")
+	fs.DurationVar(&o.LeaderElection.LeaseDuration.Duration, "leader-elect-lease-duration", o.LeaderElection.LeaseDuration.Duration, ""+
```
Comment (RainbowMango) [inline cmd/agent/app/options/options.go:101]: Seems the `o.LeaderElection.LeaseDuration.Duration` was not initialized before, which will default to `0`?

Comment (RainbowMango) [reply cmd/agent/app/options/options.go:101]: Same for the below 2 flags.

**Review by RainbowMango — COMMENTED** [review]
Maybe we have to define default values like :
https://github.com/karmada-io/karmada/blob/0074d4230eb444d0e8862eebcbbe41930f51c4ad/cmd/scheduler/app/options/options.go#L22-L24

File: cmd/agent/app/options/options.go (hunk 1 above)
Comment (CharlesQQ) [reply cmd/agent/app/options/options.go:110]: > Seems the o.LeaderElection.LeaseDuration.Duration was not initialized before?
that's right, which should be initialized in function  `NewOptions`

**Comment by CharlesQQ** [comment]
> Do you mean the default [renew deadline](https://github.com/kubernetes-sigs/controller-runtime/blob/ffd9ec8768b7d2e82aee6b0c76d481a5bdda4075/pkg/manager/manager.go#L202) (defaults to 10s) and [RetryPeriod](https://github.com/kubernetes-sigs/controller-runtime/blob/ffd9ec8768b7d2e82aee6b0c76d481a5bdda4075/pkg/manager/manager.go#L205) (defaults to 2s) might be too short for you, right?

File: cmd/agent/app/options/options.go [hunk 2]
```diff
@@ -96,8 +105,21 @@ func (o *Options) AddFlags(fs *pflag.FlagSet, allControllers []string) {
 		"A list of controllers to enable. '*' enables all on-by-default controllers, 'foo' enables the controller named 'foo', '-foo' disables the controller named 'foo'. All controllers: %s.",
 		strings.Join(allControllers, ", "),
 	))
-	fs.BoolVar(&o.LeaderElection.LeaderElect, "leader-elect", true, "Start a leader election client and gain leadership before executing the main loop. Enable this when running replicated components for high availability.")
+	fs.BoolVar(&o.LeaderElection.LeaderElect, "leader-elect", o.LeaderElection.LeaderElect, "Start a leader election client and gain leadership before executing the main loop. Enable this when running replicated components for high availability.")
```
Comment (RainbowMango) [inline cmd/agent/app/options/options.go:108]: I like the previous `true` better, it's more straightforward.

File: cmd/agent/app/options/options.go (hunk 1 above)
Comment (RainbowMango) [reply cmd/agent/app/options/options.go:110]: Please use `defaultElectionLeaseDuration` as default value, just for readability concern.

File: cmd/agent/app/options/options.go [hunk 3]
```diff
@@ -82,6 +88,9 @@ func NewOptions() *Options {
 			LeaderElect:       true,
 			ResourceLock:      resourcelock.LeasesResourceLock,
 			ResourceNamespace: util.NamespaceKarmadaSystem,
+			LeaseDuration:     defaultElectionLeaseDuration,
+			RenewDeadline:     defaultElectionRenewDeadline,
+			RetryPeriod:       defaultElectionRetryPeriod,
```
Comment (RainbowMango) [inline cmd/agent/app/options/options.go:93]: I think we don't need it since it will be set when initializing the flag.

Comment (CharlesQQ) [reply cmd/agent/app/options/options.go]: got it!

**Comment by RainbowMango** [comment]
Looks good to me now. Thanks for your quick response.
/lgtm

Can you help test it? And share a test report/result?

And, we need a release note for this PR description, for example:
```
`karmada-controller-manager`/`karmada-agent`: Introduced `--leader-elect-lease-duration`, `--leader-elect-renew-deadline` and `--leader-elect-retry-period` flags to specify leader election behavious.
```

**Comment by CharlesQQ** [comment]
> Can you help test it? And share a test report/result?

start controller with option ` --leader-elect-lease-duration=60s --leader-elect-renew-deadline=30s --leader-elect-retry-period=10s`, update leases.coordination.k8s.io object period changed:
```
kubectl get leases.coordination.k8s.io -n karmada-system karmada-controller-manager -o custom-columns='rename:.metadata.name,nenewTime:.spec.renewTime' --watch

before:
karmada-controller-manager   2022-06-25T10:32:25.990626Z
karmada-controller-manager   2022-06-25T10:32:28.023668Z
karmada-controller-manager   2022-06-25T10:32:30.034589Z
karmada-controller-manager   2022-06-25T10:32:32.046473Z
karmada-controller-manager   2022-06-25T10:32:34.063017Z
karmada-controller-manager   2022-06-25T10:32:36.074157Z

after
karmada-controller-manager   2022-06-25T11:18:57.441700Z
karmada-controller-manager   2022-06-25T11:18:57.587956Z
karmada-controller-manager   2022-06-25T11:19:07.605609Z
karmada-controller-manager   2022-06-25T11:19:17.685018Z
karmada-controller-manager   2022-06-25T11:19:27.697147Z
karmada-controller-manager   2022-06-25T11:19:37.783649Z
```

**Comment by RainbowMango** [comment]
Thanks @CharlesQQ 
/lgtm
/approve

**Pull request merged**